### PR TITLE
Issue: 5377 - Code cleanup: Fix Covscan invalid reference

### LINKED ISSUE #5377: Code cleanup: Fix Covscan invalid reference
**Issue Description**
 - Fix errors listed at https://cov01.lab.eng.brq.redhat.com/covscanhub/waiving/95389/263550/#defects
 - Jira ticket is: https://issues.redhat.com/browse/IDMDS-2394

### PR DESCRIPTION
**Pull request by progier389**
A code cleanup to get less covscan errors

Issue: [5377](https://github.com/389ds/389-ds-base/issues/5377)
Reviewer: @mreynolds389 

Initial covscan result                |   Current covscan result
|-|-|    
ARRAY_VS_SINGLETON 9           | ARRAY_VS_SINGLETON        7
BAD_COMPARE 1                  |
BAD_FREE 2                     | BAD_FREE                  2
BUFFER_SIZE 8                  | BUFFER_SIZE               3
CHECKED_RETURN 18              | CHECKED_RETURN            8
CLANG_WARNING 336              | CLANG_WARNING             248
COMPILER_WARNING 26            | COMPILER_WARNING          7
CONSTANT_EXPRESSION_RESULT 4   | CONSTANT_EXPRESSION_RESULT 2
COPY_PASTE_ERROR 8             | COPY_PASTE_ERROR          6
CPPCHECK_WARNING 38            | CPPCHECK_WARNING          29
DC.WEAK_CRYPTO 5               | DC.WEAK_CRYPTO            5
DEADCODE 22                    | DEADCODE                  4
FORWARD_NULL 47                | FORWARD_NULL              5
GCC_ANALYZER_WARNING 89        | GCC_ANALYZER_WARNING      77
|| IDENTICAL_BRANCHES        2
LOCK 4                         | LOCK                      2
MISSING_BREAK 5                | MISSING_BREAK             5
MISSING_LOCK 5                 | MISSING_LOCK              4
NEGATIVE_RETURNS 3             | NEGATIVE_RETURNS          2
NO_EFFECT 1                    | NO_EFFECT                 1
NULL_RETURNS 2                 | NULL_RETURNS              1
OVERRUN 18                     | OVERRUN                   18
RESOURCE_LEAK 29               | RESOURCE_LEAK             13
REVERSE_INULL 11               | REVERSE_INULL             1
SENSITIVE_DATA_LEAK 1          | SENSITIVE_DATA_LEAK       1
SHELLCHECK_WARNING 4           | SHELLCHECK_WARNING        4
SIZEOF_MISMATCH 6              | SIZEOF_MISMATCH           1
STRING_OVERFLOW 16             | STRING_OVERFLOW           16
TAINTED_SCALAR 4               | TAINTED_SCALAR            3
TAINTED_STRING 1               | TAINTED_STRING            1
TOCTOU 6                       | TOCTOU                    6
UNINIT 3                       | UNINIT                    1
UNREACHABLE 2                  | UNREACHABLE               3
UNUSED_VALUE 16                | UNUSED_VALUE              10
USELESS_CALL 3                 | USELESS_CALL              3
USE_AFTER_FREE 12              | USE_AFTER_FREE            8

### CODE REVIEW

File: ldap/servers/plugins/replication/repl5_agmt.c [hunk 1]
```diff
@@ -401,6 +401,10 @@ agmt_new_from_entry(Slapi_Entry *e)
             replica_incr_agmt_count(replica);
         }
     }
+    if (!replica) {
+        slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name, "agmt_new_from_entry - Failed to get replica.\n");
+        goto loser;
```
Comment (mreynolds389) [inline ldap/servers/plugins/replication/repl5_agmt.c:408]: Can we add more info about which replication agreement (dn), or some other identifying info?

File: ldap/servers/slapd/add.c [hunk 2]
```diff
@@ -818,6 +818,7 @@ op_shared_add(Slapi_PBlock *pb)
     slapi_ch_free((void **)&pwdtype);
     slapi_ch_free_string(&proxydn);
     slapi_ch_free_string(&proxystr);
+    delete_passwdPolicy(&pwpolicy);
```
Comment (mreynolds389) [inline ldap/servers/slapd/add.c]: I don't think this is needed (false positive).   new_passwdPolicy() stores the new passwordPolicy object in the pblock, and when the pblock is destroyed so is this passwordPolicy struct.

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c [hunk 3]
```diff
@@ -2116,13 +2112,24 @@ void *dbmdb_recno_cache_build(void *arg)
         dbmdb_generate_recno_cache_key_by_recno(&rckey, recno);
         rc = MDB_PUT(txn_ctx.txn, rcctx->rcdbi->dbi, &rckey, &rcdata, 0);
         slapi_ch_free(&rckey.mv_data);
-        if (rc == 0) {
+        if (rc) {
+            slapi_log_err(SLAPI_LOG_ERR, "dbmdb_recno_cache_build", 
+                          "Failed to write record in db %s, key=%s error: %s\n", 
+                            rcctx->rcdbi->dbname, (char*)(key.mv_data), mdb_strerror(rc));
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c]: indentation :)

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c [hunk 4]
```diff
@@ -2116,13 +2112,24 @@ void *dbmdb_recno_cache_build(void *arg)
...
+        } else {
             dbmdb_generate_recno_cache_key_by_data(&rckey, &key, &data);
             rc = MDB_PUT(txn_ctx.txn, rcctx->rcdbi->dbi, &rckey, &rcdata, 0);
             slapi_ch_free(&rckey.mv_data);
             txn_ctx.flags |= DBMDB_TXNCTX_NEED_COMMIT;
+            if (rc) {
+                slapi_log_err(SLAPI_LOG_ERR, "dbmdb_recno_cache_build", 
+                              "Failed to write record in db %s, key=%s error: %s\n", 
+                                rcctx->rcdbi->dbname, (char*)(key.mv_data), mdb_strerror(rc));
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c]: indentation like above

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c [hunk 5]
```diff
@@ -2116,13 +2112,24 @@ void *dbmdb_recno_cache_build(void *arg)
...
+            }
         }
         slapi_ch_free(&rcdata.mv_data);
+        if (rc) { 
+        }
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c]: Is this empty block intentional?

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_rdncache.c [hunk 6]
```diff
@@ -351,7 +350,6 @@ rdncache_new_elem(RDNcacheHead_t *head, ID entryid, ID parentid, int nrdnlen,
     }
     head->head_per_id[idx_by_id] = elem;
     head->head_per_rdn[idx_by_rdn] = elem;
-    idx_by_id = rdncache_lookup_by_id(head, parentid);
     elem->head = head;
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/db-mdb/mdb_rdncache.c:353]: Just curious, what is this fixing?  I wonder if this is related to import crash I found...

File: ldap/servers/slapd/passwd_extop.c [hunk 7]
```diff
@@ -874,6 +875,9 @@ passwd_modify_extop(Slapi_PBlock *pb)
 
     /* We can free the generated password bval now */
     ber_bvfree(gen_passwd);
+    if (pwpolicy) {
+        delete_passwdPolicy(&pwpolicy);
+    }
```
Comment (mreynolds389) [inline ldap/servers/slapd/passwd_extop.c]: False positive, passwordPolicy struct is freed when pblock is destroyed

File: ldap/servers/slapd/pw.c [hunk 8]
```diff
@@ -251,6 +252,7 @@ slapi_encode_ext(Slapi_PBlock *pb, const Slapi_DN *sdn, char *value, char *alg)
     }
 
     hashedval = (*pws_enc)(value);
+    delete_passwdPolicy(&pwpolicy);
```
Comment (mreynolds389) [inline ldap/servers/slapd/pw.c]: Not needed

File: ldap/servers/slapd/pw.c [hunk 9]
```diff
@@ -1767,6 +1769,7 @@ add_password_attrs(Slapi_PBlock *pb, Operation *op __attribute__((unused)), Slap
         slapi_entry_attr_merge(e, "passwordallowchangetime", bvals);
         slapi_ch_free_string(&bv.bv_val);
     }
+    delete_passwdPolicy(&pwpolicy);
```
Comment (mreynolds389) [inline ldap/servers/slapd/pw.c]: not needed

File: ldap/servers/slapd/pw.c [hunk 10]
```diff
@@ -2873,8 +2876,10 @@ slapi_check_account_lock(Slapi_PBlock *pb, Slapi_Entry *bind_target_entry, int p
 
 notlocked:
     /* account is not locked. */
+    delete_passwdPolicy(&pwpolicy);
     return (0);
 locked:
+    delete_passwdPolicy(&pwpolicy);
     return (1);
```
Comment (mreynolds389) [inline ldap/servers/slapd/pw.c:2888]: Don't think these are needed either.

File: ldap/servers/slapd/pw_mgmt.c [hunk 11]
```diff
@@ -116,6 +116,7 @@ need_new_pw(Slapi_PBlock *pb, Slapi_Entry *e, int pwresponse_req)
             pw_apply_mods(sdn, &smods);
         }
         slapi_mods_done(&smods);
+        delete_passwdPolicy(&pwpolicy);
```
Comment (mreynolds389) [inline ldap/servers/slapd/pw_mgmt.c]: Same here

**Comment by mreynolds389** [comment]
I stopped leaving comment for all the "delete password policy" calls.  They are all false positives.

As for cargo.lock changes.  I recently updated this file and merged it.  Perhaps this change to the file should be removed from this PR?

Besides my minor comments the rest looks good.

**Review by mreynolds389 — DISMISSED** [review]
Looks good, but unless cargo.lock was updated to fix an error I think it should be removed from this PR.

**Comment by mreynolds389** [comment]
Ok a majority of CI tests are crashing the server.  Something must be off with this PR.  Please investigate.

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_rdncache.c (hunk 6 above)
Comment (progier389) [reply ldap/servers/slapd/back-ldbm/db-mdb/mdb_rdncache.c:353]: I do not think iy is related to the crash.  It fix a variable set but not used warning ( idx_by_id is not used afterwards so there is no need to compute it ... )

File: ldap/servers/plugins/replication/repl5_agmt.c (hunk 1 above)
Comment (progier389) [reply ldap/servers/plugins/replication/repl5_agmt.c:408]: Ok: I am changing it to:
        slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name, 
                      "agmt_new_from_entry - Failed to get replica for agreement %s on replicated suffix %s).\n",
                      slapi_entry_get_dn(e), tmpstr ? tmpstr : "<NULL>");

File: ldap/servers/slapd/add.c (hunk 2 above)
Comment (progier389) [reply ldap/servers/slapd/add.c]: I did not realized that new_passwdPolicy was registering the policy in the pblock,
 so I will have to double check all delete_passwdPolicy change I have done and use /* coverity[leaked_storage] */ instead.

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c (hunk 4 above)
Comment (progier389) [reply ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c]: :+1:

**Comment by progier389** [comment]
> g the server. Something must be off with this PR. Please investigate.

IMHO that is likely related to the pwpolicy double free let see if next commit does fix the issue ...

File: ldap/servers/slapd/add.c (hunk 2 above)
Comment (mreynolds389) [reply ldap/servers/slapd/add.c]: Little history on this: the original work was done in:  https://github.com/389ds/389-ds-base/issues/147  Basically in every single operation we would allocate the password policy struct like 3-4 times :-)  So we ended up storing it in the pblock to save from doing all these duplicate allocations.

File: ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c (hunk 5 above)
Comment (progier389) [reply ldap/servers/slapd/back-ldbm/db-mdb/mdb_layer.c]: No. I wanted to add to check on code return but that is not even the right place ... Fixing that ...

**Comment by progier389** [comment]
Waiting for the complete test but there are still lots of errors ...

**Comment by progier389** [comment]
Tests are much better but there are still some suspicious failure that I will have to investigate on tomorrow.

**Comment by progier389** [comment]
This time the tes are OK ...

File: ldap/servers/slapd/back-ldbm/cache.c [hunk 12]
```diff
@@ -2220,8 +2225,16 @@ cache_is_in_cache(struct cache *cache, void *ptr)
         return in_cache;
     }
     bep = (struct backcommon *)ptr;
-    cache_lock(cache);
     in_cache = (bep->ep_state & (ENTRY_STATE_DELETED | ENTRY_STATE_NOTINCACHE)) ? 0 : 1;
-    cache_unlock(cache);
     return in_cache;
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/cache.c:2229]: "cache" is no longer needed in this function (unused parameter).  Perhaps we can refactor this function?

**Comment by mreynolds389** [comment]
I still think the Cargo.lock changes should be removed, the reset looks good except for one very minor issue (unused parameter)

**Review by mreynolds389 — APPROVED** [review]

**Comment by progier389** [comment]
And removed the Cargo file

File: ldap/servers/slapd/back-ldbm/cache.c (hunk 12 above)
Comment (progier389) [reply ldap/servers/slapd/back-ldbm/cache.c:2229]: Ok, I refacteored cache_is_in_cache_nolock in last update.

Comment (mreynolds389) [reply ldap/servers/slapd/back-ldbm/cache.c:2229]: thanks, approval has been set!

**Pull request merged**